Loader implementation of `icon` metadata field does not match spec

From Fabric wiki, the [mod metadata specification](https://fabricmc.net/wiki/documentation:fabric_mod_json_spec#optional_fields_metadata) says this about icons:
>  **icon**: Contains the mod's icon, as a square .PNG file. (Minecraft resource packs use 128×128, but that is not a hard requirement - a power of two is, however, recommended.) Can be provided in one of two forms:
> 
 >   A string, providing the path (from the mod's root) to a single .PNG file,
  >  A string→string dictionary, where the keys conform to widths of each PNG file, and the values are said files' paths.

However, Loader fails to load a mod with a metadata file specifying multiple icons:

```
Caused by: net.fabricmc.loader.impl.metadata.ParseMetadataException: Error reading fabric.mod.json file for mod at C:\Users\abby\AppData\Roaming\PrismLauncher\instances\adventure-platform-fabric test\.minecraft\mods\adventure-platform-fabric-5.5.1-SNAPSHOT.jar: Icon path must be a string Error was located at:  at line 19 column 6 path $.icon.
	at net.fabricmc.loader.impl.metadata.V1ModMetadataParser.readIcon(V1ModMetadataParser.java:599)
	at net.fabricmc.loader.impl.metadata.V1ModMetadataParser.parse(V1ModMetadataParser.java:196)
	at net.fabricmc.loader.impl.metadata.ModMetadataParser.readModMetadata(ModMetadataParser.java:142)
	at net.fabricmc.loader.impl.metadata.ModMetadataParser.readModMetadata(ModMetadataParser.java:105)
	at net.fabricmc.loader.impl.metadata.ModMetadataParser.parseMetadata(ModMetadataParser.java:45)
	at net.fabricmc.loader.impl.discovery.ModDiscoverer$ModScanTask.parseMetadata(ModDiscoverer.java:469)
	at net.fabricmc.loader.impl.discovery.ModDiscoverer$ModScanTask.computeJarFile(ModDiscoverer.java:291)
	at net.fabricmc.loader.impl.discovery.ModDiscoverer$ModScanTask.compute(ModDiscoverer.java:253)
	at net.fabricmc.loader.impl.discovery.ModDiscoverer$ModScanTask.compute(ModDiscoverer.java:212)
	at java.base/java.util.concurrent.RecursiveTask.exec(RecursiveTask.java:100)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
```

The mod metadata file in question had the following data:

```yaml
icon:
  128: assets/adventure/logo.png
  64: assets/adventure/logo-64px.png
```

This disconnect between the spec and the implementation is confusing for myself as a mod developer, and adds unnecessary effort for tooling authors who might write libraries to read Fabric mod metadata.

--- a/src/main/java/net/fabricmc/loader/impl/metadata/V1ModMetadataParser.java
+++ b/src/main/java/net/fabricmc/loader/impl/metadata/V1ModMetadataParser.java
@@ -595,10 +595,6 @@ final class V1ModMetadataParser {
 			final SortedMap<Integer, String> iconMap = new TreeMap<>(Comparator.naturalOrder());
 
 			while (reader.hasNext()) {
-				if (reader.peek() != JsonToken.STRING) {
-					throw new ParseMetadataException("Icon path must be a string", reader);
-				}
-
 				String key = reader.nextName();
 
 				int size;
@@ -612,6 +608,12 @@ final class V1ModMetadataParser {
 				if (size < 1) {
 					throw new ParseMetadataException("Size must be positive!", reader);
 				}
+
+				if (reader.peek() != JsonToken.STRING) {
+					throw new ParseMetadataException("Icon path must be a string", reader);
+				}
+
+				iconMap.put(size, reader.nextString());
 			}
 
 			reader.endObject();

--- a/src/test/java/net/fabricmc/test/V1ModJsonParsingTests.java
+++ b/src/test/java/net/fabricmc/test/V1ModJsonParsingTests.java
@@ -29,6 +29,7 @@ import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.Collections;
 import java.util.Map;
+import java.util.Optional;
 
 import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.DisplayName;
@@ -156,6 +157,42 @@ final class V1ModJsonParsingTests {
 		if (modMetadata.getDependencies().isEmpty()) {
 			throw new RuntimeException("Incorrect amount of dependencies");
 		}
+	}
+
+	@Test
+	@DisplayName("Any icon size test file")
+	public void testAnyIconSizeFile() throws IOException, ParseMetadataException {
+		final LoaderModMetadata metadata = parseMetadata(specPath.resolve("any_icon_size.json"));
+
+		validateIconPath(metadata, 32, 64);
+		validateIconPath(metadata, 64, 64);
+		validateIconPath(metadata, 128, 64);
+	}
+
+	@Test
+	@DisplayName("No icon test file")
+	public void testNoIconFile() throws IOException, ParseMetadataException {
+		final LoaderModMetadata metadata = parseMetadata(specPath.resolve("no_icon.json"));
+		assertEquals(Optional.empty(), metadata.getIconPath(32));
+	}
+
+	@Test
+	@DisplayName("Icon sizes test file")
+	public void testIconSizesFile() throws IOException, ParseMetadataException {
+		final LoaderModMetadata metadata = parseMetadata(specPath.resolve("icon_sizes.json"));
+
+		validateIconPath(metadata, 32, 64);
+		validateIconPath(metadata, 64, 64);
+		validateIconPath(metadata, 128, 128);
+		validateIconPath(metadata, 256, 256);
+		validateIconPath(metadata, 512, 256);
+	}
+
+	private void validateIconPath(LoaderModMetadata metadata, int preferredSize, int expectedSize) {
+		Optional<String> expected = Optional.of("assets/testing/icon-" + expectedSize + ".png");
+		Optional<String> actual = metadata.getIconPath(preferredSize);
+
+		assertEquals(expected, actual, "Found " + actual.orElse("no icon") + " for preferred size " + preferredSize + " instead of expected size " + expectedSize);
 	}
 
 	/*

--- a/src/test/resources/testing/parsing/v1/spec/any_icon_size.json
+++ b/src/test/resources/testing/parsing/v1/spec/any_icon_size.json
@@ -1,0 +1,6 @@
+{
+  "schemaVersion": 1,
+  "id": "v1-parsing-test",
+  "version": "1.0.0-SNAPSHOT",
+  "icon": "assets/testing/icon-64.png"
+}

--- a/src/test/resources/testing/parsing/v1/spec/icon_sizes.json
+++ b/src/test/resources/testing/parsing/v1/spec/icon_sizes.json
@@ -1,0 +1,10 @@
+{
+  "schemaVersion": 1,
+  "id": "v1-parsing-test",
+  "version": "1.0.0-SNAPSHOT",
+  "icon": {
+    "64": "assets/testing/icon-64.png",
+    "128": "assets/testing/icon-128.png",
+    "256": "assets/testing/icon-256.png"
+  }
+}

--- a/src/test/resources/testing/parsing/v1/spec/no_icon.json
+++ b/src/test/resources/testing/parsing/v1/spec/no_icon.json
@@ -1,0 +1,5 @@
+{
+  "schemaVersion": 1,
+  "id": "v1-parsing-test",
+  "version": "1.0.0-SNAPSHOT"
+}

